src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,147 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> previous;
+
+        public Node(Node<T> previous, T value, Node<T> next) {
+            this.value = value;
+            this.next = next;
+            this.previous = previous;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            head.previous = newNode;
+            newNode.next = head;
+            head = newNode;
+            size++;
+        } else {
+            checkIndex(index);
+            Node<T> currentNode = getNodeByIndex(index);
+            newNode.next = currentNode;
+            newNode.previous = currentNode.previous;
+            currentNode.previous.next = newNode;
+            currentNode.previous = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (int index = 0; index < list.size(); index++) {
+                add(list.get(index));
+            }
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T valueToReturn = getNodeByIndex(index).value;
```

Comment:
Here, you're calling `getNodeByIndex(index) `twice. This is not efficient; consider storing the node in a variable to avoid the double lookup